File: playwright/S3.js
import "dotenv/config";
import fs from "fs";
import path from "path";
import {
  GetObjectCommand,
  ListObjectsCommand,
  PutObjectCommand,
  S3Client,
} from "@aws-sdk/client-s3";

const {
  AWS_BUCKET_NAME: bucketName,
  AWS_BUCKET_REGION: region,
  AWS_ACCESS_KEY: accessKeyId,
  AWS_SECRET_ACCESS_KEY: secretAccessKeyId,
} = process.env;

const S3_Client = new S3Client({
  credentials: {
    accessKeyId: accessKeyId,
    secretAccessKey: secretAccessKeyId,
  },
  region: region,
});

export async function list_objects() {
  console.log(`bucket name ::: ${bucketName} - region ::: ${region} ::: ${accessKeyId} ::: ${secretAccessKeyId}`)
  const params = {
    Bucket: 'doctor-octopus',
    Prefix: "test_reports/"
  };

  const command = new ListObjectsCommand(params);
  const response = await S3_Client.send(command);
  // console.log(`List S3 response ::: ${JSON.stringify(response.Contents)}`);

  return response;
}

export async function get_object(imageName) {
  const params = {
    Bucket: bucketName,
    Key: imageName,
  };

  const command = new GetObjectCommand(params);
  const response = await S3_Client.send(command);
  // console.log(`Get S3 response ::: ${JSON.stringify(response)}`);
  return response;
}

export async function upload_file(bucket_name, key, file_content) {
  const params = {
    Bucket: bucket_name,
    Key: key,
    Body: file_content
  };

  const command = new PutObjectCommand(params);
  return await S3_Client.send(command);
}

export const upload_directory = async (bucket_name, local_dir_path, s3_dir_path) => {
  console.log(`Uploading directory ::: '${local_dir_path}' to S3 bucket ::: ${bucket_name} > bucket directory ::: ${s3_dir_path}`);
  const report_dir = fs.readdirSync(local_dir_path);

  for (const r_item of report_dir) {
    const r_item_path = path.join(local_dir_path, r_item);
    const stats = fs.statSync(r_item_path);

    if(stats.isFile()) {
      const file_content = fs.readFileSync(r_item_path, 'utf-8');
      await upload_file(bucket_name, `${s3_dir_path}/${r_item}`, file_content);
    } else if(stats.isDirectory()) {
      await upload_directory(bucket_name, r_item_path, `${s3_dir_path}/${r_item}`);
    }
  }
}

File: server/src/util/s3.py
from dotenv import load_dotenv
load_dotenv('.env')
import os
import boto3

aws_bucket_name = os.environ.get('AWS_BUCKET_NAME')
aws_bucket_region = os.environ.get('AWS_BUCKET_REGION')
aws_access_key = os.environ.get('AWS_ACCESS_KEY')
aws_secret_access_key = os.environ.get('AWS_SECRET_ACCESS_KEY')

# S3 client
class S3Client:
    def __init__(self):
        session = boto3.Session(
            aws_access_key_id=aws_access_key,
            aws_secret_access_key=aws_secret_access_key,
            region_name=aws_bucket_region
        )
        self.S3 = session.client('s3')

    def get_a_s3_object(self, object_name, bucket_name = aws_bucket_name):
        '''
        # Call S3 clientto get a s3 object
        '''
        response = self.S3.get_object(Bucket=bucket_name, Key=object_name)
        return response['Body'].read()
    
    def list_s3_objects(self, bucket_name = aws_bucket_name):
        '''
        # Call S3 client to list s3 objects
        '''
        response = self.S3.list_objects_v2(Bucket=bucket_name)
        objects = response.get('Contents', [])
        # for obj in objects:
        #     print(f"Key: {obj['Key']}, Size: {obj['Size']}")
        return objects
    
S3 = S3Client()
